cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll

```diff
@@ -0,0 +1,519 @@
+  strictcount (op.getOperator()) = 1
+}
+
+private predicate mk_UnaryOp(HC child, string opname, UnaryOperation op) {
+  analyzableUnaryOp(op) and
+  child = hashCons(op.getOperand().getFullyConverted()) and
+  opname = op.getOperator()
+}
+
+private predicate analyzableThisExpr(ThisExpr thisExpr) {
+  strictcount(thisExpr.getEnclosingFunction()) = 1
+}
+
+private predicate mk_ThisExpr(Function fcn, ThisExpr thisExpr) {
+  analyzableThisExpr(thisExpr) and
+  fcn = thisExpr.getEnclosingFunction()
+}
+
+private predicate analyzableArrayAccess(ArrayExpr ae) {
+  strictcount (ae.getArrayBase().getFullyConverted()) = 1 and
+  strictcount (ae.getArrayOffset().getFullyConverted()) = 1
+}
+
+private predicate mk_ArrayAccess(
+  HC base, HC offset, ArrayExpr ae) {
+  analyzableArrayAccess(ae) and
+  base = hashCons(ae.getArrayBase().getFullyConverted()) and
+  offset = hashCons(ae.getArrayOffset().getFullyConverted())
+}
+
+private predicate analyzablePointerDereferenceExpr(
+  PointerDereferenceExpr deref) {
+  strictcount (deref.getOperand().getFullyConverted()) = 1
+}
+
+private predicate mk_Deref(
+  HC p, PointerDereferenceExpr deref) {
+  analyzablePointerDereferenceExpr(deref) and
+  p = hashCons(deref.getOperand().getFullyConverted())
+}
+
+private predicate analyzableNonmemberFunctionCall(
+  FunctionCall fc) {
+  forall(int i | exists(fc.getArgument(i)) | strictcount(fc.getArgument(i)) = 1) and
+  strictcount(fc.getTarget()) = 1 and
+  not fc.getTarget().isMember()
+}
+
+private predicate mk_NonmemberFunctionCall(
+  Function fcn,
+  HC_Args args,
+  FunctionCall fc
+) {
+  fc.getTarget() = fcn and
+  analyzableNonmemberFunctionCall(fc) and
+  (
+    exists(HC head, HC_Args tail |
+      args = HC_ArgCons(fcn, head, fc.getNumberOfArguments() - 1, tail) and
+      mk_ArgCons(fcn, head, fc.getNumberOfArguments() - 1, tail, fc)
+    )
+    or
+    fc.getNumberOfArguments() = 0 and
+    args = HC_EmptyArgs(fcn)
+  )
+}
+
+private predicate analyzableMemberFunctionCall(
+  FunctionCall fc) {
+  forall(int i | exists(fc.getArgument(i)) | strictcount(fc.getArgument(i)) = 1) and
+  strictcount(fc.getTarget()) = 1 and
+  strictcount(fc.getQualifier()) = 1
+}
+
+private predicate mk_MemberFunctionCall(
+  Function fcn,
+  HC qual,
+  HC_Args args,
+  FunctionCall fc
+) {
+  fc.getTarget() = fcn and
+  analyzableMemberFunctionCall(fc) and
+  hashCons(fc.getQualifier()) = qual and
+  (
+    exists(HC head, HC_Args tail |
+      args = HC_ArgCons(fcn, head, fc.getNumberOfArguments() - 1, tail) and
+      mk_ArgCons(fcn, head, fc.getNumberOfArguments() - 1, tail, fc)
+    )
+    or
+    fc.getNumberOfArguments() = 0 and
+    args = HC_EmptyArgs(fcn)
+  )
+}
+
+private predicate analyzableFunctionCall(
+  FunctionCall fc
+) {
+  analyzableNonmemberFunctionCall(fc)
+  or
+  analyzableMemberFunctionCall(fc)
+}
+
+/**
+ * Holds if `fc` is a call to `fcn`, `fc`'s first `i-1` arguments have hash-cons
+ * `list`, and `fc`'s `i`th argument has hash-cons `hc`
```

Comment:
I think this comment is off by one. Where it says `i-1` it should just say `i`. Take, for example, the case of `i=0`. Then it should say that "`fc`'s first **0** arguments have hash-cons `list`, and `fc`'s **0th** argument has hash-cons `hc`.

Maybe also change "`fc`'s `i`th argument" to "`fc`'s argument at index `i`" or clarify in some other way that the first argument has `i=0`. Also add a full stop at the end of the sentence.

Reply:
done